spring-cloud-netflix-eureka-client/src/main/java/org/springframework/cloud/netflix/eureka/http/WebClientEurekaHttpClient.java

```diff
@@ -0,0 +1,236 @@
+				.header(HttpHeaders.CONTENT_TYPE, MediaType.APPLICATION_JSON_VALUE)
+				.header(HttpHeaders.ACCEPT, MediaType.APPLICATION_JSON_VALUE).exchange()
+				.block();
+
+		int statusCode = statusCodeValueOf(response);
+
+		Applications body = response.toEntity(Applications.class).block().getBody();
+
+		return anEurekaHttpResponse(statusCode,
+				statusCode == HttpStatus.OK.value() && body != null ? body : null)
+				.headers(headersOf(response)).build();
+	}
+
+	@Override
+	public EurekaHttpResponse<Applications> getDelta(String... regions) {
+		return getApplicationsInternal("apps/delta", regions);
+	}
+
+	@Override
+	public EurekaHttpResponse<Applications> getVip(String vipAddress, String... regions) {
+		return getApplicationsInternal("vips/" + vipAddress, regions);
+	}
+
+	@Override
+	public EurekaHttpResponse<Applications> getSecureVip(String secureVipAddress,
+			String... regions) {
+		return getApplicationsInternal("svips/" + secureVipAddress, regions);
+	}
+
+	@Override
+	public EurekaHttpResponse<Application> getApplication(String appName) {
+
+		ClientResponse response = webClient.get()
+				.uri("apps/" + appName, Application.class)
+				.header(HttpHeaders.ACCEPT, MediaType.APPLICATION_JSON_VALUE).exchange()
+				.block();
+
+		int statusCode = statusCodeValueOf(response);
+		Application body = response.toEntity(Application.class).block().getBody();
+
+		Application application = statusCode == HttpStatus.OK.value() && body != null
+				? body : null;
+
+		return anEurekaHttpResponse(statusCode, application).headers(headersOf(response))
+				.build();
+	}
+
+	@Override
+	public EurekaHttpResponse<InstanceInfo> getInstance(String appName, String id) {
+		return getInstanceInternal("apps/" + appName + '/' + id);
+	}
+
+	@Override
+	public EurekaHttpResponse<InstanceInfo> getInstance(String id) {
+		return getInstanceInternal("instances/" + id);
+	}
+
+	private EurekaHttpResponse<InstanceInfo> getInstanceInternal(String urlPath) {
+		ClientResponse response = webClient.get().uri(urlPath, InstanceInfo.class)
+				.header(HttpHeaders.ACCEPT, MediaType.APPLICATION_JSON_VALUE).exchange()
+				.block();
+
+		int statusCode = statusCodeValueOf(response);
+		InstanceInfo body = response.toEntity(InstanceInfo.class).block().getBody();
+
+		return anEurekaHttpResponse(statusCode,
+				statusCode == HttpStatus.OK.value() && body != null ? body : null)
+				.headers(headersOf(response)).build();
+	}
+
+	@Override
+	public void shutdown() {
+		// Nothing to do
+	}
+
+	private static Map<String, String> headersOf(ClientResponse response) {
+		ClientResponse.Headers httpHeaders = response.headers();
+		if (httpHeaders == null) {
+			return Collections.emptyMap();
+		}
+		HttpHeaders asHeaders = httpHeaders.asHttpHeaders();
+		if (asHeaders == null) {
+			return Collections.emptyMap();
+		}
+		Map<String, String> headers = new HashMap<>();
+		asHeaders.entrySet().stream().forEach(entry -> entry.getValue().stream()
```

Comment:
The double stream.forEach() could probably be replaced by using `Map.forEach()`?